### DISCUSSION TITLE
Loading images from the file system

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#12994] [Q&A]
[_From discord_:](https://discord.com/channels/372137812037730304/550416151172087808/1133013360292679773)

Hi all. I am new to the Platform. I work for a company using UNO and I have a project that I need to show some images from a folder. I have done the XAML and the data structures. It  works when I select to run on Windows platform but whe I use GTK (Linux) the images do not show. I have checked file paths and they seems ok. I use the Path.Combine to handle Cross platform paths correctly. My xaml is this 

```
        <ListView Grid.Column="0"  ItemsSource="{Binding Procedure.ComponentsList}" SelectionMode="None">

            <ListView.ItemsPanel>
                <ItemsPanelTemplate>
                    <ItemsStackPanel  VerticalAlignment="Top"/>
                </ItemsPanelTemplate>
            </ListView.ItemsPanel>

            <ListView.ItemTemplate>
                <DataTemplate>
                    <Grid>
                        <Grid.ColumnDefinitions>
                            <ColumnDefinition Width=""/>
                            <ColumnDefinition Width=""/>
                        </Grid.ColumnDefinitions>

                        <TextBlock Grid.Column="0" HorizontalAlignment="Left" VerticalAlignment="Center" TextAlignment="Left" Text="{Binding CompName}" Style="{StaticResource TextBlockBlue}" FontSize="40"/>
                        <Image Grid.Column="1" Stretch="Uniform" Source="{Binding ImageIcon}"  />
                    </Grid>
                </DataTemplate>
            </ListView.ItemTemplate>
        </ListView>
```
        
[8:30 AM]Cellerate: Does anybody have an idea what that may be?
[8:31 AM]Cellerate: The binding value is this   <Image Grid.Column="1" Stretch="Uniform" Source="{Binding ImageIcon}"  />
[9:16 AM]Cellerate: The imageIcon when i bind it to a TextBlock I can see the path like that... <TextBlock Grid.Column="0" HorizontalAlignment="Left" VerticalAlignment="Center" TextAlignment="Left" Text="{Binding ImageIcon}" Style="{StaticResource TextBlockBlue}" FontSize="40"/>
[9:19 AM]Cellerate: The Image is not local instead the path is absolute.. like this /mnt/c/Users/Harry/Desktop/AssemblerV4.0/process/procedures/Cellerate_A/icons/ImageName.png
[9:19 AM]Cellerate: I use WSL to run in Linux
[9:21 AM]Cellerate: Any ideas?

**Comment by jeromelaban** [comment]
Thanks for the question. In general, image loading errors are reported in the app logs. Can you show your logs? Otherwise, local file system loading may need a special path, or maybe it's a scenarion uno does not support explicitly.

If you have a repro of this, can you attach it here?

**Reply by hnCellerate** [reply]
Hi Jerom,

Thank you for your response. I appreciate it.

 I have not seen in the logs anything about Image loading fail message. Also I don't see anything in the output window of Visual Studio 2022. I also use Serilog. Is there anything I have to do to log it there?

I have made a test code to verify...

        this.procedure.ImageIcon2 = new BitmapImage();
        var impath = Path.Combine( FileOperations.RootPath,"process", "procedures", "Cellerate_A", "Icons", "CoinCellBase.png");
        var uriSource = new Uri(impath);

        this.procedure.ImageIcon2.ImageFailed += (sender, e) => {
            Debug.WriteLine("Image Failed");
        };
        this.procedure.ImageIcon2.UriSource = uriSource;

The XAML is this
<Image Stretch="Uniform" Source="{Binding Procedure.ImageIcon2}"  />


When I run this code in Windows  I can see the image on the screen. To Test this code I change the image  "CoinCellBase.png" to  "CoinCellBase.Spng" and then the ImageFailed event gets activated. So I know that the file was not found...

But when I compile and run in GTK (WSL2 Ubuntu) (I have also tested on real hardware) and the image file name is correct The image is not showing on the screen and when I put wrong filename the Event ImageFailed **is NOT activated**.

This is the URI I get from debugging for Linux (WSL2).  
AbsolutePath = "file:///mnt/c/Users/Harry/Desktop/AssemblerV4.0/process/procedures/Cellerate_A/Icons/CoinCellBase.png" 

and this is for Windows.
AbsolutePath = "file:///C:/Users/Harry/Desktop/AssemblerV4.0/process/procedures/Cellerate_A/Icons/CoinCellBase.prng"

When I deploy the App  on the real hardware and I copy the path generated by the code and do from terminal a
ls -la /home/cellerate/harry/process/procedures/Cellerate_A/Icons/ 
it does show me the images stored there so the path MUST be right...


I was able to use images with relative path from the Assets folder for another screen though BUT I need to load images from a folder outside the App folder...

I tend to believe this is an UNO bug (since the ImageFailed event is not activated) for Linux but then again it might be that needs special handling for Linux targets....


Kind regards 
Harry

**Reply by hnCellerate** [reply]
Sorry forgot to paste XAML...
<Image Stretch="Uniform" Source="{Binding Procedure.ImageIcon2}"  />

**Reply by jeromelaban** [reply]
Thanks for the update. This is actually quite interesting and it must be a new behavior of WinUI that was not present in UWP. Previously, opening local files could only be done through `BitmapImage.SetSource`, or through `ms-appdata`, but not using `file:///` scheme.

Let's track this as a new feature in https://github.com/unoplatform/uno/issues/13003

**Reply by hnCellerate** [reply]
Thank you Jerome. 

In the mean time I have to devise a workaround... I know that I can get to show an Image on GTK from the internet with a URL (hhtps://...). I was wondering if I can load a local file as a URL address...

I there a release roadmap URL page that I can see when your next update will be available?


Kind Regards
Harry

**Reply by jeromelaban** [reply]
You can use `BitmapImage.SetSource` to achieve the same effect. We generally do not provide timeline on updates, but if you're willing to work on adding support for it and make a PR in Uno, the community can help.

**Comment by hnCellerate** [comment]
`<Image Stretch="Uniform" Source="{Binding Procedure.ImageIcon2}"  />`
